source/common/http/http2/codec_impl.cc

```diff
@@ -1107,104 +1176,50 @@ Status ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
   // and CONTINUATION frames in onBeforeFrameReceived().
   ASSERT(frame->hd.type != NGHTTP2_CONTINUATION);
 
-  if ((frame->hd.type == NGHTTP2_PING) && (frame->ping.hd.flags & NGHTTP2_FLAG_ACK)) {
+  if (frame->hd.type == NGHTTP2_PING) {
     // The ``opaque_data`` should be exactly what was sent in the ping, which is
     // was the current time when the ping was sent. This can be useful while debugging
     // to match the ping and ack.
     uint64_t data;
     safeMemcpy(&data, &(frame->ping.opaque_data));
-    ENVOY_CONN_LOG(trace, "recv PING ACK {}", connection_, data);
-
-    onKeepaliveResponse();
-    return okStatus();
+    return onPing(data, frame->ping.hd.flags & NGHTTP2_FLAG_ACK);
   }
 
   if (frame->hd.type == NGHTTP2_DATA) {
-    RETURN_IF_ERROR(trackInboundFrames(frame->hd.stream_id, frame->hd.length, frame->hd.type,
-                                       frame->hd.flags, frame->data.padlen));
+    return onBeginData(frame->hd.stream_id, frame->hd.length, frame->hd.type, frame->hd.flags,
+                       frame->data.padlen);
   }
 
   // Only raise GOAWAY once, since we don't currently expose stream information. Shutdown
   // notifications are the same as a normal GOAWAY.
   // TODO: handle multiple GOAWAY frames.
-  if (frame->hd.type == NGHTTP2_GOAWAY && !raised_goaway_) {
-    ASSERT(frame->hd.stream_id == 0);
-    raised_goaway_ = true;
-    callbacks().onGoAway(ngHttp2ErrorCodeToErrorCode(frame->goaway.error_code));
-    return okStatus();
+  if (frame->hd.type == NGHTTP2_GOAWAY) {
+    return onGoAway(frame->goaway.error_code);
   }
 
   if (frame->hd.type == NGHTTP2_SETTINGS && frame->hd.flags == NGHTTP2_FLAG_NONE) {
     std::vector<http2::adapter::Http2Setting> settings;
+    settings.reserve(frame->settings.niv);
     for (const nghttp2_settings_entry& entry :
          absl::MakeSpan(frame->settings.iv, frame->settings.niv)) {
       settings.push_back(
           {static_cast<http2::adapter::Http2SettingsId>(entry.settings_id), entry.value});
     }
     onSettings(settings);
+    return okStatus();
   }
 
   StreamImpl* stream = getStreamUnchecked(frame->hd.stream_id);
   if (!stream) {
     return okStatus();
   }
 
-  // Track bytes sent and received.
-  if (frame->hd.type != METADATA_FRAME_TYPE) {
-    stream->bytes_meter_->addWireBytesReceived(frame->hd.length + H2_FRAME_HEADER_SIZE);
-  }
-  if (frame->hd.type == NGHTTP2_HEADERS || frame->hd.type == NGHTTP2_CONTINUATION) {
-    stream->bytes_meter_->addHeaderBytesReceived(frame->hd.length + H2_FRAME_HEADER_SIZE);
+  if (frame->hd.type == NGHTTP2_HEADERS) {
+    return onHeaders(frame->hd.stream_id, frame->hd.length, frame->hd.flags, frame->headers.cat);
   }
```

Comment:
Should there still be a call to addWireBytesReceived() and addHeaderBytesReceived() for NGHTTP2_CONTINUATION?

edit: Ah, I see https://nghttp2.org/documentation/enums.html#c.nghttp2_frame_type.NGHTTP2_CONTINUATION - this should not be in any callbacks IIUC.

Reply:
Yes, this was already dead code, in that nghttp2 does not call the `on_frame_recv_callback` for CONTINUATION frames.

This does make me wonder how header frame wire bytes are accounted for, as the HEADERS frame header can't possibly include future CONTINUATION frame bytes.